api/scrutiny: return non empty results

jordipainan opened the pull request:
A bit hacky until state computed results

jordipainan committed:
api/scrutiny: return non empty results, a bit hacky until state computed results

app.GetProcess returns all the information about a process, including its results.
The results array returned is an array already initialized with length equal
to the number of oracles * 2 (this is done due consensus reasons).
For serving the results via API we do not want to return void results and
we cannot consider them valid.
On top of this, oracles can report different results. We agreed on, and
until results are computed as part of the state, comparing all the non void
results and only return them if they are equal.

This commit fixes returning void results by checking if an oracle
actually signed them and checks that all the oracles reported the same
results. So if the results are not void and the same we return them
as valid via API. Otherwise an error is returned.

mvdan commented on api/elections.go:
```diff
@@ -281,58 +281,84 @@ func (a *API) electionScrutinyHandler(msg *apirest.APIdata, ctx *httprouter.HTTP
+			log.Warnf("different number of votes")
+			return false
+		}
+		for k, questionResult := range otherResult.Votes {
```
to be consistent, you could name the parameters `result1, result2`, and then the votes `vote1, vote2` at the top of this loop. You're indexing `Votes[k]` in multiple places otherwise.

mvdan commented on api/elections.go:
```diff
@@ -281,58 +281,84 @@ func (a *API) electionScrutinyHandler(msg *apirest.APIdata, ctx *httprouter.HTTP
 				}
-				if len(questionResult.Question) != len(firstResult.Votes[k].Question) {
-					log.Errorf("question length do not match")
+				if len(baseResult.Votes[k].Question[kk]) != len(questionOption) {
```
FYI: bytes.Equal already checks lengths, so this is in theory not necessary

jordipainan replied:
yup!

mvdan reviewed (approved):
Perhaps explain in the commit message why it's wanted? but in any case the code seems fine to me.

p4u commented on api/elections.go:
```diff
@@ -281,58 +281,82 @@ func (a *API) electionScrutinyHandler(msg *apirest.APIdata, ctx *httprouter.HTTP
+	equalResults := func(baseResult, otherResult *models.ProcessResult) bool {
+		// if the number of votes is different, the results are different
+		if len(baseResult.Votes) != len(otherResult.Votes) {
+			log.Warnf("different number of votes")
```
I think its good to print some more debug info and to use the new Warnw() function for this

jordipainan replied:
Yup, I've changed to Warnw() and I've added more useful information

p4u commented on api/elections.go:
```diff
@@ -281,58 +281,82 @@ func (a *API) electionScrutinyHandler(msg *apirest.APIdata, ctx *httprouter.HTTP
+	if len(intermediateResults) == 0 {
+		return fmt.Errorf("no results available")
+	}
+	firstResult := intermediateResults[0]
```
This assignment should go next to the place where it is used, so line 342

jordipainan replied:
Right! Changed

p4u commented on api/elections.go:
```diff
@@ -281,58 +281,82 @@ func (a *API) electionScrutinyHandler(msg *apirest.APIdata, ctx *httprouter.HTTP
+	} // at this point results are equal
+
+	// cast results to big.Int
+	results := make([][]*big.Int, len(firstResult.Votes))
```
don't we have a function somewhere that already does that?

jordipainan replied:
No AFAIK, but I've created a new one just below the `GetFriendlyResults` so now we can reuse it when is needed :)

jordipainan committed:
apply suggestions

p4u commented on vochain/state/utils.go:
```diff
@@ -31,6 +31,18 @@ func GetFriendlyResults(results []*models.QuestionResult) [][]string {
 }
 
+// ResultsToBigIntMatrix converts the results of a process to a big.Int matrix.
+func ResultsToBigIntMatrix(results []*models.QuestionResult) [][]*big.Int {
```
We are repeating code. What about modifying GetFriendlyResults to return a `[][]*types.bigInt` instead of [][]string ?

When types.bigInt is stringified or json-marshaled, it correctly converts the contented to a String.

jordipainan replied:
Made the proper changes for enabling this changes in the last commit :)

jordipainan committed:
GetFriendlyResults and API results now work with types.BigInt:

In the past we were using strings in API results responses
and in the GetFriendlyResults state method, but given that the
wrapped bigInt is able to serialize properly when wire encoding
is required we can just use our own BigInt type.

For abiEncoded results is still required to use the standard
math/big because upstream code does not allow to work directly
with custom defined types.

Also the types.BigInt.ToInt() method is renamed to
ToStdBigInt() cause the old naming is very confusing because
someone expect to get an Int but gets instead a math/big type.

p4u commented on api/censuses.go:
```diff
@@ -201,7 +201,7 @@ func (a *API) censusAddHandler(msg *apirest.APIdata, ctx *httprouter.HTTPContext
 			p.Weight = new(types.BigInt).SetUint64(1)
 		}
-		if ref.Indexed && p.Weight.ToInt().Uint64() != 1 {
+		if ref.Indexed && p.Weight.ToStdBigInt().Uint64() != 1 {
```
What's the point for replacing "ToInt()"  to "ToStdBigInt" ?

jordipainan replied:
When I was coding this PR, specifically the part touching the abi.Encode I needed a method to cast results from types.BigInt to math/big.
When I looked at the code at first glance I did not find anything and later on, I found that ToInt()  what it actually does is to cast to math/big. This is, at least, anti-intuitive and I cannot find a single reason on why someone should expect that a method called ToInt() instead of casting a bigInt to Int cast from types.BigInt to a math/big BigInt.

So, I've decided to change the name to a more suited one.

p4u replied:
I see. So `ToInt()` method name did not provide enough info about what it actually does. However if we are changing this, I'd not use`ToStdBigInt`, taking ideas from other big.Int wrappers, a better naming IMO would be `MathBigInt()` which clearly indicates what is the method returning.

jordipainan replied:
I agree that MathBigInt() is a better naming. 
When I'm searching for methods, my brain would expect ToMathBigInt() but is just opinionated I guess, the next thing I would do is to search without the To keyword so I guess MathBigInt() is fine too :)
Thank you for the suggestion on the naming.

jordipainan replied:
Changed :)

p4u commented on rpcclient/api.go:
```diff
@@ -246,11 +246,11 @@ func (c *Client) TestResults(pid []byte, totalVotes int, withWeight uint64) (res
-		results[1][2] != total ||
-		results[2][3] != total ||
-		results[3][4] != total {
+	total := uint64(totalVotes) * withWeight
```
You can make total a big.Int here so you don't need to declare new variables on each comparison.

jordipainan replied:
yup! changed

p4u commented on types/big.go:
```diff
@@ -23,11 +23,11 @@ func (i *BigInt) UnmarshalText(data []byte) error {
 
 func (i *BigInt) GobEncode() ([]byte, error) {
-	return i.ToInt().GobEncode()
+	return i.ToStdBigInt().GobEncode()
```
This naming change is out of the scope for this PR :(

jordipainan replied:
Well, the scope of this PR was to quickly fix the API retuning empty results, and we've spent 1 week iterating over some things that are not in the scope of this PR. 
I agree that this naming change could be included in another PR, but given how this PR evolved, the change does not hurt anyone and IMHO can be included here.
At this point, I prefer not to roll back the changes.

p4u replied:
Ok. But a separated commit with this change would have been appreciated.

jordipainan replied:
Gotcha, I will do it better next time.

p4u commented on api/elections.go:
```diff
@@ -281,66 +281,104 @@ func (a *API) electionScrutinyHandler(msg *apirest.APIdata, ctx *httprouter.HTTP
+			for kk, question1 := range vote1.Question {
+				question2 := vote2.Question[kk]
+				if question1 == nil {
+					log.Warnw("invalid question option (nil)",
```
Please rebase with master. The syntax for using log key/value functions has changed, now it expects a list of key/values instead of a map (this way is more standard and comfortable). The correct way to use this functions would be something like: `log.Warnw("invalid question", "question", nil, "index", kk)`

p4u commented on api/elections.go:
```diff
@@ -281,66 +281,104 @@ func (a *API) electionScrutinyHandler(msg *apirest.APIdata, ctx *httprouter.HTTP
+				if !bytes.Equal(question1, question2) {
+					log.Warnw("question option bytes do not match",
+						map[string]interface{}{
+							"baseResult question option":  fmt.Sprintf("%x", question2),
```
So, there is no need to make so explicit the name of the value. On the first parameter you are adding the context message and the next just print useful variable information in key/value format.:

`"question option do not match", "option1", fmt.Sprintf("%x", question2), "option2", fmt.Sprintf("%x", question1))`

p4u commented on api/helpers.go:
```diff
@@ -115,7 +116,15 @@ func encodeEVMResultsArgs(electionId common.Hash,
+			resultsStd[i][j] = v.ToStdBigInt()
+		}
+	}
+	abiEncodedResultsBytes, err := args.Pack(electionId, organizationId, censusRoot, sourceContractAddress, resultsStd)
```
@mvdan is this loop required for transforming types.BigInt to big.Int? Isn't a way to make a direct casting? (just wondering)

mvdan replied:
unfortunately the loop is required right now. even though the types are the same in memory, they are different named types in the Go type system, so you can't go from one slice to another directly. we could probably avoid this awkwardness with either generics or avoiding the need for `types.BigInt` to exist at all, but probably for another PR. perhaps raise an issue and I can look at it when we're done with sqlite :)

p4u commented on oracle/oracle.go:
```diff
@@ -153,11 +153,20 @@ func (o *Oracle) OnComputeResults(results *indexertypes.Results, proc *indexerty
+			resultsString[i][j] = w.String()
+		}
+	}
 	signedResultsPayload := OracleResults{
```
OracleResults.Results could be of type `[][]*types.BigInt` since JSON marhals it as String.

jordipainan committed:
apply suggestions
